### ISSUE TITLE
Piwik JS Tracker: Easier way for using multiple trackers

### ISSUE THREAD

**Issue by tsteur** [piwik/piwik#9525]
For example in https://github.com/piwik/piwik/issues/9518 we had a case where two scripts declare a Piwik and overwrite each other. In http://developer.piwik.org/guides/tracking-javascript-guide#multiple-piwik-trackers we explain how to use multiple trackers eg by calling

```js
var piwikTracker = Piwik.getTracker(12,  "http://example.com/piwik/" );
piwikTracker.trackPageView();
```

However, creating the tracker instance is not trivial and is a bit different from the standard async tracker. For example a developer would need to make sure to call methods on the tracker in this order to configure the tracker:

```js
['disableCookies', 'setTrackerUrl', 'setAPIUrl', 'setCookiePath', 'setCookieDomain', 'setDomains', 'setUserId', 'setSiteId', 'enableLinkTracking'];
```

Otherwise there might be side effects. Also one would have to change all calls to Piwik tracker from say `_paq.push(['trackPageView])` to `if (piwikTracker) piwikTracker.trackPageView();` etc. which can be a lot of effort.

I wonder if there are easier ways. Eg could we maybe also provide a `_paq2`, `_paq3`, `_paq4`, ...? 
We probably don't want to pollute the namespace so maybe we could do something like `_paq2  =_paq.create()` or `paq2 = _paqCreate();`. This way one could create unlimited `_paq` instances manually as needed. All a developer has to do afterwards is to replace `_paq` with `_paq2` and one is done. The order of the applied methods that are applied would be the same as usual etc.

**Comment by RMastop** [comment]
That would be an awesome improvement! Would both trackers register downloads, outlinks, etc?

**Comment by tsteur** [comment]
Yes I'd say so but I'm not sure if there are any problems with it. I think it should work fine though

**Comment by meghashriKirani** [comment]
So this is working is it?
How can I get the namespace in my tracker, for example, paq2 = _paqCreate(); to a particular tracker if I have 2 trackers. An example would be very helpful.

**Comment by tsteur** [comment]
@meghashriKirani have a look here: http://developer.piwik.org/guides/tracking-javascript-guide#multiple-piwik-trackers

**Comment by mattab** [comment]
PR in progress: https://github.com/piwik/piwik/pull/10085

**Comment by tsteur** [comment]
BTW is not 100% going to be solved by #10085. #10085 is only a solution if someone wants to 100% track the same data to both Piwik instances. If someone wants to track different data to different Piwiks a different solution is needed.

**Comment by meghashriKirani** [comment]
Thanks.. Please keep me posted.

On 21 April 2016 at 06:34, Thomas Steur <notifications@github.com> wrote:

> BTW is not 100% going to be solved by #10085
> <https://github.com/piwik/piwik/pull/10085>. #10085
> <https://github.com/piwik/piwik/pull/10085> is only a solution if someone
> wants to 100% track the same data to both Piwik instances. If someone wants
> to track different data to different Piwiks a different solution is needed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/piwik/piwik/issues/9525#issuecomment-212738647>
>



-- 
Best regards,

*Meghashri Kirani*
Technical analyst

ABOUT YOU GmbH
Domstraße 10
20095 Hamburg

**Comment by sebastianpiskorski** [comment]
@tsteur I would consider also a scenario where somebody uses two Piwik.js scripts from different versions. There is namespace conflict in such situation because all is stored in global `Piwik` object.
